exercises/react/src/example/kotlin/React.kt

```diff
@@ -0,0 +1,72 @@
+class Reactor<T>() {
+    abstract inner class Cell {
+        abstract val value: T
+        internal val dependents = mutableListOf<ComputeCell>()
+        internal fun eachDependent(f: (ComputeCell) -> Any) {
+            for (dep in dependents) {
+                f(dep)
+            }
+        }
+    }
+
+    interface Subscription {
+        fun cancel()
+    }
+
+    inner class InputCell(initialValue: T) : Cell() {
+        override var value: T = initialValue
+            set(newValue) {
+                field = newValue
+                eachDependent { it.propagate() }
+                eachDependent { it.fireCallbacks() }
+            }
+    }
+
+    inner class ComputeCell private constructor(val newValue: () -> T) : Cell() {
+        override var value: T = newValue()
+            private set
+
+        private var lastCallbackValue = value
+        private var callbacksIssued = 0
+        private val activeCallbacks = mutableMapOf<Int, (T) -> Any>()
+
+        constructor(c: Cell, f: (T) -> T) : this({ f(c.value) }) {
+            c.dependents.add(this)
+        }
+        constructor(c1: Cell, c2: Cell, f: (T, T) -> T) : this({ f(c1.value, c2.value) }) {
+            c1.dependents.add(this)
+            c2.dependents.add(this)
+        }
+
+        fun addCallback(f: (T) -> Any): Subscription {
```

Comment:
Sigh, I changed too many intervening lines, so the old comment is on an "outdated diff" now. Gotta post it again.

I hope this is not too strange of an API. The API looks something like:

```kotlin
val cell: ComputeCell = getAComputeCell();
val sub = cell.addCallback { print(it) }
// ...
sub.cancel()
```

The alternative is to have `addCallback` return some sort of ID:


```kotlin
val cell: ComputeCell = getAComputeCell();
val subID = cell.addCallback { print(it) }
// ...
cell.removeCallback(subID)
```